mockgcp/mockspanner/admin/instance.go

```diff
@@ -0,0 +1,150 @@
+	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/spanner/admin/instance/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/emptypb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+var _ pb.InstanceAdminServer = &SpannerInstanceV1{}
+
+type SpannerInstanceV1 struct {
+	*MockService
+	pb.UnimplementedInstanceAdminServer
+}
+
+func (s *SpannerInstanceV1) GetInstance(ctx context.Context, req *pb.GetInstanceRequest) (*pb.Instance, error) {
+	name, err := s.parseInstanceName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+	fqn := name.String()
+
+	obj := &pb.Instance{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if status.Code(err) == codes.NotFound {
+			return nil, status.Errorf(codes.NotFound, "Instance not found: %s", name.String())
+		}
+		return nil, err
+	}
+	return obj, nil
+}
+
+func (s *SpannerInstanceV1) CreateInstance(ctx context.Context, req *pb.CreateInstanceRequest) (*longrunningpb.Operation, error) {
+	instanceName := req.GetParent() + "/instances/" + req.GetInstanceId()
+	name, err := s.parseInstanceName(instanceName)
+	if err != nil {
+		return nil, err
+	}
+	fqn := name.String()
+	now := timestamppb.Now()
+
+	obj := proto.Clone(req.GetInstance()).(*pb.Instance)
+	obj.Name = fqn
+	s.populateDefaultsForSpannerInstance(obj, obj)
+	obj.State = pb.Instance_READY
+	obj.CreateTime = now
+	obj.UpdateTime = now
+	if err := s.storage.Create(ctx, fqn, obj); err != nil {
+		return nil, err
+	}
+	metadata := &pb.CreateInstanceMetadata{
+		Instance:                  obj,
+		StartTime:                 now,
+		EndTime:                   now,
+		ExpectedFulfillmentPeriod: pb.FulfillmentPeriod_FULFILLMENT_PERIOD_NORMAL,
+	}
+	return s.operations.DoneLRO(ctx, name.String(), metadata, obj)
+}
+
+func (s *SpannerInstanceV1) populateDefaultsForSpannerInstance(update, obj *pb.Instance) {
+	// At most one of either node_count or processing_units should be present.
+	// https://cloud.google.com/spanner/docs/compute-capacity
+	// 1 nodeCount equals to 1000 processingUnits
+	if 1000*update.NodeCount > update.ProcessingUnits {
+		obj.ProcessingUnits = 1000 * update.NodeCount
+		obj.NodeCount = update.NodeCount
+	} else {
+		obj.ProcessingUnits = update.ProcessingUnits
+		obj.NodeCount = update.ProcessingUnits / 1000
+	}
+}
+
+func (s *SpannerInstanceV1) UpdateInstance(ctx context.Context, req *pb.UpdateInstanceRequest) (*longrunningpb.Operation, error) {
+	name, err := s.parseInstanceName(req.Instance.Name)
+	if err != nil {
+		return nil, err
+	}
+	fqn := name.String()
+	obj := &pb.Instance{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		return nil, err
+	}
+	now := timestamppb.Now()
+	obj.UpdateTime = now
+	source := reflect.ValueOf(req.Instance)
+	target := reflect.ValueOf(obj).Elem()
+	for _, path := range req.FieldMask.Paths {
```

Comment:
A simple approach to make UPDATE more like a real GCP UPDATE with update mask_field.

Reply:
Nice!  We should split this out into a helper method!